File: src/praisonai/praisonai/cli/features/sandbox_executor.py
"""
Sandboxed Execution System for PraisonAI CLI.

Inspired by Codex CLI's sandbox modes for secure command execution.
Provides isolated execution environment when enabled via --sandbox flag.

Architecture:
- SandboxExecutor: Base class for sandboxed execution
- SubprocessSandbox: Subprocess-based isolation with resource limits
- SandboxPolicy: Configurable security policies

Note: Sandbox is ONLY activated when explicitly requested via CLI flag.
"""

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Set
from pathlib import Path
from enum import Enum
import subprocess
import logging
import tempfile
import shutil
import time

logger = logging.getLogger(__name__)


# ============================================================================
# Sandbox Modes
# ============================================================================

class SandboxMode(Enum):
    """
    Sandbox execution modes.
    
    - DISABLED: No sandboxing (default)
    - BASIC: Basic isolation with resource limits
    - STRICT: Strict isolation with filesystem restrictions
    - NETWORK_ISOLATED: No network access
    """
    DISABLED = "disabled"
    BASIC = "basic"
    STRICT = "strict"
    NETWORK_ISOLATED = "network_isolated"
    
    @classmethod
    def from_string(cls, value: str) -> "SandboxMode":
        """Parse mode from string."""
        value = value.lower().strip().replace("-", "_")
        for mode in cls:
            if mode.value == value:
                return mode
        raise ValueError(f"Unknown sandbox mode: {value}")


# ============================================================================
# Data Classes
# ============================================================================

@dataclass
class SandboxPolicy:
    """
    Security policy for sandbox execution.
    """
    mode: SandboxMode = SandboxMode.DISABLED
    
    # Resource limits
    max_memory_mb: int = 512
    max_cpu_seconds: int = 30
    max_file_size_mb: int = 10
    max_processes: int = 10
    
    # Filesystem restrictions
    allowed_paths: Set[str] = field(default_factory=set)
    blocked_paths: Set[str] = field(default_factory=lambda: {
        "/etc", "/var", "/usr", "/bin", "/sbin",
        "/root", "/home", "/sys", "/proc"
    })
    temp_dir: Optional[str] = None
    
    # Network restrictions
    allow_network: bool = True
    allowed_hosts: Set[str] = field(default_factory=set)
    
    # Command restrictions
    blocked_commands: Set[str] = field(default_factory=lambda: {
        "rm", "rmdir", "mv", "dd", "mkfs", "fdisk",
        "sudo", "su", "chmod", "chown", "kill", "pkill"
    })
    
    @classmethod
    def for_mode(cls, mode: SandboxMode) -> "SandboxPolicy":
        """Create policy for a given mode."""
        if mode == SandboxMode.DISABLED:
            return cls(mode=mode)
        
        elif mode == SandboxMode.BASIC:
            return cls(
                mode=mode,
                max_memory_mb=512,
                max_cpu_seconds=60,
                allow_network=True
            )
        
        elif mode == SandboxMode.STRICT:
            return cls(
                mode=mode,
                max_memory_mb=256,
                max_cpu_seconds=30,
                max_processes=5,
                allow_network=True,
                blocked_commands={
                    "rm", "rmdir", "mv", "dd", "mkfs", "fdisk",
                    "sudo", "su", "chmod", "chown", "kill", "pkill",
                    "curl", "wget", "nc", "netcat", "ssh", "scp"
                }
            )
        
        elif mode == SandboxMode.NETWORK_ISOLATED:
            return cls(
                mode=mode,
                max_memory_mb=512,
                max_cpu_seconds=60,
                allow_network=False
            )
        
        return cls(mode=mode)


@dataclass
class ExecutionResult:
    """Result of sandboxed execution."""
    success: bool
    exit_code: int
    stdout: str
    stderr: str
    duration_ms: float
    was_sandboxed: bool
    policy_violations: List[str] = field(default_factory=list)
    
    @property
    def output(self) -> str:
        """Get combined output."""
        return self.stdout + self.stderr


# ============================================================================
# Command Validator
# ============================================================================

class CommandValidator:
    """
    Validates commands against sandbox policy.
    """
    
    def __init__(self, policy: SandboxPolicy):
        self.policy = policy
    
    def validate(self, command: str) -> List[str]:
        """
        Validate a command against the policy.
        
        Returns:
            List of policy violations (empty if valid)
        """
        violations = []
        
        if self.policy.mode == SandboxMode.DISABLED:
            return violations
        
        # Parse command
        parts = command.split()
        if not parts:
            return violations
        
        cmd_name = Path(parts[0]).name
        
        # Check blocked commands
        if cmd_name in self.policy.blocked_commands:
            violations.append(f"Command '{cmd_name}' is blocked by sandbox policy")
        
        # Check for dangerous patterns
        dangerous_patterns = [
            ("rm -rf", "Recursive force delete is blocked"),
            ("> /dev/", "Writing to device files is blocked"),
            ("| sh", "Piping to shell is blocked"),
            ("| bash", "Piping to bash is blocked"),
            ("$(", "Command substitution is blocked"),
            ("`", "Backtick command substitution is blocked"),
        ]
        
        for pattern, message in dangerous_patterns:
            if pattern in command:
                violations.append(message)
        
        # Check path access
        for part in parts:
            if part.startswith("/"):
                for blocked in self.policy.blocked_paths:
                    if part.startswith(blocked):
                        violations.append(f"Access to '{blocked}' is blocked")
                        break
        
        return violations
    
    def is_allowed(self, command: str) -> bool:
        """Check if command is allowed."""
        return len(self.validate(command)) == 0


# ============================================================================
# Subprocess Sandbox
# ============================================================================

class SubprocessSandbox:
    """
    Subprocess-based sandbox execution.
    
    Provides basic isolation through:
    - Resource limits (timeout, memory via ulimit)
    - Working directory isolation
    - Environment variable control
    - Command validation
    """
    
    def __init__(
        self,
        policy: Optional[SandboxPolicy] = None,
        working_dir: Optional[str] = None,
        verbose: bool = False
    ):
        self.policy = policy or SandboxPolicy()
        self.working_dir = Path(working_dir) if working_dir else Path.cwd()
        self.verbose = verbose
        self.validator = CommandValidator(self.policy)
        
        self._temp_dir: Optional[Path] = None
    
    def _setup_temp_dir(self) -> Path:
        """Set up temporary directory for isolated execution."""
        if self._temp_dir is None:
            self._temp_dir = Path(tempfile.mkdtemp(prefix="praisonai_sandbox_"))
        return self._temp_dir
    
    def _cleanup_temp_dir(self) -> None:
        """Clean up temporary directory."""
        if self._temp_dir and self._temp_dir.exists():
            try:
                shutil.rmtree(self._temp_dir)
            except Exception as e:
                logger.debug(f"Failed to cleanup temp dir: {e}")
            self._temp_dir = None
    
    def _build_environment(self) -> Dict[str, str]:
        """Build restricted environment variables."""
        import os
        
        # Start with minimal environment
        env = {
            "PATH": "/usr/local/bin:/usr/bin:/bin",
            "HOME": str(self._temp_dir or self.working_dir),
            "TERM": "xterm",
            "LANG": "en_US.UTF-8",
        }
        
        # Add some safe variables from current environment
        safe_vars = ["USER", "SHELL", "PYTHONPATH"]
        for var in safe_vars:
            if var in os.environ:
                env[var] = os.environ[var]
        
        return env
    
    def execute(
        self,
        command: str,
        timeout: Optional[float] = None,
        capture_output: bool = True
    ) -> ExecutionResult:
        """
        Execute a command in the sandbox.
        
        Args:
            command: Command to execute
            timeout: Timeout in seconds (overrides policy)
            capture_output: Whether to capture stdout/stderr
            
        Returns:
            ExecutionResult with execution details
        """
        start_time = time.time()
        
        # Check if sandbox is enabled
        if self.policy.mode == SandboxMode.DISABLED:
            return self._execute_unsandboxed(command, timeout, capture_output)
        
        # Validate command
        violations = self.validator.validate(command)
        if violations:
            return ExecutionResult(
                success=False,
                exit_code=-1,
                stdout="",
                stderr="Command blocked by sandbox policy:\n" + "\n".join(violations),
                duration_ms=(time.time() - start_time) * 1000,
                was_sandboxed=True,
                policy_violations=violations
            )
        
        # Set up sandbox
        if self.policy.mode == SandboxMode.STRICT:
            self._setup_temp_dir()
            cwd = self._temp_dir
        else:
            cwd = self.working_dir
        
        # Build environment
        env = self._build_environment()
        
        # Set timeout
        if timeout is None:
            timeout = float(self.policy.max_cpu_seconds)
        
        # Execute
        try:
            result = subprocess.run(
                command,
                shell=True,
                cwd=cwd,
                env=env,
                capture_output=capture_output,
                text=True,
                timeout=timeout
            )
            
            return ExecutionResult(
                success=result.returncode == 0,
                exit_code=result.returncode,
                stdout=result.stdout if capture_output else "",
                stderr=result.stderr if capture_output else "",
                duration_ms=(time.time() - start_time) * 1000,
                was_sandboxed=True
            )
        
        except subprocess.TimeoutExpired:
            return ExecutionResult(
                success=False,
                exit_code=-1,
                stdout="",
                stderr=f"Command timed out after {timeout}s",
                duration_ms=(time.time() - start_time) * 1000,
                was_sandboxed=True,
                policy_violations=["Timeout exceeded"]
            )
        
        except Exception as e:
            return ExecutionResult(
                success=False,
                exit_code=-1,
                stdout="",
                stderr=str(e),
                duration_ms=(time.time() - start_time) * 1000,
                was_sandboxed=True,
                policy_violations=[str(e)]
            )
        
        finally:
            if self.policy.mode == SandboxMode.STRICT:
                self._cleanup_temp_dir()
    
    def _execute_unsandboxed(
        self,
        command: str,
        timeout: Optional[float],
        capture_output: bool
    ) -> ExecutionResult:
        """Execute without sandbox (when disabled)."""
        start_time = time.time()
        
        try:
            result = subprocess.run(
                command,
                shell=True,
                cwd=self.working_dir,
                capture_output=capture_output,
                text=True,
                timeout=timeout
            )
            
            return ExecutionResult(
                success=result.returncode == 0,
                exit_code=result.returncode,
                stdout=result.stdout if capture_output else "",
                stderr=result.stderr if capture_output else "",
                duration_ms=(time.time() - start_time) * 1000,
                was_sandboxed=False
            )
        
        except subprocess.TimeoutExpired:
            return ExecutionResult(
                success=False,
                exit_code=-1,
                stdout="",
                stderr=f"Command timed out after {timeout}s",
                duration_ms=(time.time() - start_time) * 1000,
                was_sandboxed=False
            )
        
        except Exception as e:
            return ExecutionResult(
                success=False,
                exit_code=-1,
                stdout="",
                stderr=str(e),
                duration_ms=(time.time() - start_time) * 1000,
                was_sandboxed=False
            )


# ============================================================================
# CLI Integration Handler
# ============================================================================

class SandboxExecutorHandler:
    """
    Handler for integrating Sandbox Execution with PraisonAI CLI.
    
    Note: Sandbox is ONLY activated when --sandbox flag is passed.
    """
    
    def __init__(self, verbose: bool = False):
        self.verbose = verbose
        self._sandbox: Optional[SubprocessSandbox] = None
        self._enabled: bool = False
    
    @property
    def feature_name(self) -> str:
        return "sandbox_executor"
    
    @property
    def is_enabled(self) -> bool:
        """Check if sandbox is enabled."""
        return self._enabled
    
    def initialize(
        self,
        mode: str = "disabled",
        working_dir: Optional[str] = None,
        policy: Optional[SandboxPolicy] = None
    ) -> SubprocessSandbox:
        """
        Initialize the sandbox.
        
        Args:
            mode: Sandbox mode (disabled, basic, strict, network_isolated)
            working_dir: Working directory for execution
            policy: Optional custom policy
            
        Returns:
            Configured SubprocessSandbox
        """
        try:
            sandbox_mode = SandboxMode.from_string(mode)
        except ValueError:
            logger.warning(f"Unknown sandbox mode '{mode}', defaulting to 'disabled'")
            sandbox_mode = SandboxMode.DISABLED
        
        if policy is None:
            policy = SandboxPolicy.for_mode(sandbox_mode)
        
        self._sandbox = SubprocessSandbox(
            policy=policy,
            working_dir=working_dir,
            verbose=self.verbose
        )
        
        self._enabled = sandbox_mode != SandboxMode.DISABLED
        
        if self.verbose and self._enabled:
            from rich import print as rprint
            rprint(f"[yellow]⚠ Sandbox enabled: {sandbox_mode.value}[/yellow]")
        
        return self._sandbox
    
    def get_sandbox(self) -> Optional[SubprocessSandbox]:
        """Get the sandbox executor."""
        return self._sandbox
    
    def execute(
        self,
        command: str,
        timeout: Optional[float] = None
    ) -> ExecutionResult:
        """
        Execute a command.
        
        If sandbox is not initialized, runs without sandboxing.
        """
        if not self._sandbox:
            self._sandbox = self.initialize()
        
        return self._sandbox.execute(command, timeout=timeout)
    
    def validate_command(self, command: str) -> List[str]:
        """Validate a command against sandbox policy."""
        if not self._sandbox:
            return []
        
        return self._sandbox.validator.validate(command)
    
    def get_mode(self) -> str:
        """Get current sandbox mode."""
        if self._sandbox:
            return self._sandbox.policy.mode.value
        return SandboxMode.DISABLED.value

File: src/praisonai-ts/src/cli/features/sandbox-executor.ts
/**
 * Sandbox Executor - Safe command execution with restrictions
 */

export type SandboxMode = 'disabled' | 'basic' | 'strict' | 'network-isolated';

export interface SandboxConfig {
  mode: SandboxMode;
  allowedCommands?: string[];
  blockedCommands?: string[];
  blockedPaths?: string[];
  timeout?: number;
  maxOutputSize?: number;
  cwd?: string;
  env?: Record<string, string>;
}

export interface ExecutionResult {
  success: boolean;
  stdout: string;
  stderr: string;
  exitCode: number;
  duration: number;
  truncated?: boolean;
}

/**
 * Default blocked commands (dangerous operations)
 */
export const DEFAULT_BLOCKED_COMMANDS = [
  'rm -rf /',
  'rm -rf /*',
  'rm -rf ~',
  'rm -rf ~/*',
  'mkfs',
  'dd if=/dev/zero',
  'dd if=/dev/random',
  ':(){ :|:& };:',
  'chmod -R 777 /',
  'chown -R',
  'sudo rm',
  'sudo dd',
  'sudo mkfs',
  'shutdown',
  'reboot',
  'halt',
  'poweroff',
  'init 0',
  'init 6'
];

/**
 * Default blocked paths
 */
export const DEFAULT_BLOCKED_PATHS = [
  '/etc/passwd',
  '/etc/shadow',
  '/etc/sudoers',
  '/root',
  '~/.ssh',
  '~/.gnupg',
  '~/.aws',
  '~/.config/gcloud',
  '/var/log',
  '/boot',
  '/sys',
  '/proc'
];

/**
 * Command validator
 */
export class CommandValidator {
  private blockedCommands: string[];
  private blockedPaths: string[];
  private allowedCommands?: string[];

  constructor(config: Partial<SandboxConfig> = {}) {
    this.blockedCommands = config.blockedCommands || DEFAULT_BLOCKED_COMMANDS;
    this.blockedPaths = config.blockedPaths || DEFAULT_BLOCKED_PATHS;
    this.allowedCommands = config.allowedCommands;
  }

  /**
   * Validate a command
   */
  validate(command: string): { valid: boolean; reason?: string } {
    const normalized = command.toLowerCase().trim();

    // Check allowlist first if specified
    if (this.allowedCommands) {
      const baseCmd = normalized.split(/\s+/)[0];
      if (!this.allowedCommands.includes(baseCmd)) {
        return { valid: false, reason: `Command '${baseCmd}' not in allowlist` };
      }
    }

    // Check blocked commands
    for (const blocked of this.blockedCommands) {
      if (normalized.includes(blocked.toLowerCase())) {
        return { valid: false, reason: `Blocked command pattern: ${blocked}` };
      }
    }

    // Check blocked paths
    for (const blockedPath of this.blockedPaths) {
      const expandedPath = blockedPath.replace('~', process.env.HOME || '');
      if (normalized.includes(expandedPath.toLowerCase())) {
        return { valid: false, reason: `Blocked path: ${blockedPath}` };
      }
    }

    // Check for shell injection patterns
    const dangerousPatterns = [
      /;\s*rm\s/i,
      /\|\s*rm\s/i,
      /`.*`/,
      /\$\(.*\)/,
      />\s*\/dev\/sd/i,
      />\s*\/dev\/null.*2>&1.*&/i
    ];

    for (const pattern of dangerousPatterns) {
      if (pattern.test(command)) {
        return { valid: false, reason: 'Potentially dangerous pattern detected' };
      }
    }

    return { valid: true };
  }
}

/**
 * Sandbox Executor class
 */
export class SandboxExecutor {
  private config: SandboxConfig;
  private validator: CommandValidator;

  constructor(config: Partial<SandboxConfig> = {}) {
    this.config = {
      mode: 'basic',
      timeout: 30000,
      maxOutputSize: 1024 * 1024, // 1MB
      cwd: process.cwd(),
      ...config
    };
    this.validator = new CommandValidator(this.config);
  }

  /**
   * Execute a command in the sandbox
   */
  async execute(command: string): Promise<ExecutionResult> {
    const startTime = Date.now();

    // Disabled mode - no execution
    if (this.config.mode === 'disabled') {
      return {
        success: false,
        stdout: '',
        stderr: 'Sandbox execution is disabled',
        exitCode: 1,
        duration: 0
      };
    }

    // Validate command
    const validation = this.validator.validate(command);
    if (!validation.valid) {
      return {
        success: false,
        stdout: '',
        stderr: `Command rejected: ${validation.reason}`,
        exitCode: 1,
        duration: 0
      };
    }

    try {
      const result = await this.spawn(command);
      return {
        ...result,
        duration: Date.now() - startTime
      };
    } catch (error: any) {
      return {
        success: false,
        stdout: '',
        stderr: error.message,
        exitCode: 1,
        duration: Date.now() - startTime
      };
    }
  }

  /**
   * Spawn the command
   */
  private async spawn(command: string): Promise<Omit<ExecutionResult, 'duration'>> {
    const { spawn } = await import('child_process');

    return new Promise((resolve) => {
      const env = this.buildEnv();
      
      const proc = spawn('sh', ['-c', command], {
        cwd: this.config.cwd,
        env,
        timeout: this.config.timeout,
        stdio: ['pipe', 'pipe', 'pipe']
      });

      let stdout = '';
      let stderr = '';
      let truncated = false;
      const maxSize = this.config.maxOutputSize || 1024 * 1024;

      proc.stdout?.on('data', (data) => {
        if (stdout.length < maxSize) {
          stdout += data.toString();
          if (stdout.length > maxSize) {
            stdout = stdout.slice(0, maxSize);
            truncated = true;
          }
        }
      });

      proc.stderr?.on('data', (data) => {
        if (stderr.length < maxSize) {
          stderr += data.toString();
          if (stderr.length > maxSize) {
            stderr = stderr.slice(0, maxSize);
            truncated = true;
          }
        }
      });

      proc.on('close', (code) => {
        resolve({
          success: code === 0,
          stdout,
          stderr,
          exitCode: code || 0,
          truncated
        });
      });

      proc.on('error', (error) => {
        resolve({
          success: false,
          stdout,
          stderr: error.message,
          exitCode: 1
        });
      });
    });
  }

  /**
   * Build environment variables based on sandbox mode
   */
  private buildEnv(): Record<string, string> {
    const baseEnv = { ...process.env, ...this.config.env };

    switch (this.config.mode) {
      case 'strict':
        // Minimal environment
        return {
          PATH: '/usr/local/bin:/usr/bin:/bin',
          HOME: process.env.HOME || '',
          USER: process.env.USER || '',
          SHELL: '/bin/sh',
          TERM: 'dumb',
          ...this.config.env
        };

      case 'network-isolated':
        // No network access (requires additional OS-level setup)
        return {
          ...baseEnv,
          http_proxy: 'http://localhost:0',
          https_proxy: 'http://localhost:0',
          HTTP_PROXY: 'http://localhost:0',
          HTTPS_PROXY: 'http://localhost:0',
          no_proxy: '',
          NO_PROXY: ''
        };

      case 'basic':
      default:
        return baseEnv as Record<string, string>;
    }
  }

  /**
   * Check if a command would be allowed
   */
  wouldAllow(command: string): { allowed: boolean; reason?: string } {
    if (this.config.mode === 'disabled') {
      return { allowed: false, reason: 'Sandbox execution is disabled' };
    }
    const validation = this.validator.validate(command);
    return { allowed: validation.valid, reason: validation.reason };
  }

  /**
   * Get current sandbox mode
   */
  getMode(): SandboxMode {
    return this.config.mode;
  }

  /**
   * Update sandbox mode
   */
  setMode(mode: SandboxMode): void {
    this.config.mode = mode;
  }
}

/**
 * Create a sandbox executor
 */
export function createSandboxExecutor(config?: Partial<SandboxConfig>): SandboxExecutor {
  return new SandboxExecutor(config);
}

/**
 * Quick execute with default sandbox
 */
export async function sandboxExec(command: string, config?: Partial<SandboxConfig>): Promise<ExecutionResult> {
  const executor = createSandboxExecutor(config);
  return executor.execute(command);
}
